### PR TITLE
Laboratory work #4, Damir Khamidullin - 22FPL2

### PR DESCRIPTION
**Pull request by Damir-Khamidullin**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +31,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
+        for symbol in ['.', '!', '?']:
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:36]: use tuple in this case

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +31,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+            text = text.replace(symbol, ' ' + self._end_of_word_token)
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:37]: use fstring

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -39,6 +58,10 @@ def _put(self, element: str) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(element, str) or not element:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:62]: add message to the error

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -28,6 +31,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:35]: add message

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -28,6 +31,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+        for symbol in ['.', '!', '?']:
+            text = text.replace(symbol, ' ' + self._end_of_word_token)
+        tokens = []
+        for word in text.lower().split():
+            if word == self._end_of_word_token or word.isalpha():
+                tokens.append(word)
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:41]: Is it some optimization?

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -56,7 +79,13 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
-
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:83]: message

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -56,7 +79,13 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
...
+        corpus = ' '.join(decoded_corpus).split(self._end_of_word_token)
+        text = '. '.join([sent.strip().capitalize() for sent in corpus])
+        if text[-1] == ' ':
+            return text[:-1]
+        return text + '.'
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:88]: fstring

File: lab_4_fill_words_by_ngrams/main.py [hunk 8]
```diff
@@ -98,6 +130,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if (not isinstance(seq_len, int) or seq_len < 1 or
+           not isinstance(prompt, str) or not prompt):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:135]: message

File: lab_4_fill_words_by_ngrams/main.py [hunk 9]
```diff
@@ -98,6 +130,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        prompt_encoded = self._word_processor.encode(prompt)
+        if not prompt_encoded:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:138]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 10]
```diff
@@ -98,6 +130,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        while seq_len > 0:
+            tokens = self._model.generate_next_token(prompt_encoded)
+            if tokens is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:142]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 11]
```diff
@@ -98,6 +130,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            tokens_to_choose_from = []
+            for token in tokens_sorted:
+                if freq_sum < self._p_value:
+                    freq_sum += token[1]
+                    tokens_to_choose_from.append(token[0])
+            prompt_encoded += (random.choice(tokens_to_choose_from),)
+            seq_len -= 1
+        text = self._word_processor.decode(prompt_encoded)
+        if not text:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:157]: .

**Review by artyomtugaryov — COMMENTED** [review]
Add error messages in each raise

File: lab_4_fill_words_by_ngrams/main.py [hunk 12]
```diff
@@ -220,6 +295,25 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
                 or if methods used return None,
                 or if nothing was generated.
         """
+        if not isinstance(generated_text, str) or not generated_text:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:299]: message

File: lab_4_fill_words_by_ngrams/main.py [hunk 13]
```diff
@@ -220,6 +295,25 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
...
+        text_encoded = self._word_processor.encode(generated_text)
+        if not text_encoded:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:302]: message

File: lab_4_fill_words_by_ngrams/main.py [hunk 14]
```diff
@@ -220,6 +295,25 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
...
+        ngram_size = self._language_model.get_n_gram_size()
+        logs = 0.
+        for i in range(ngram_size - 1, len(text_encoded)):
+            context = tuple(text_encoded[i - ngram_size + 1: i])
+            tokens = self._language_model.generate_next_token(context)
+            if not tokens:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:309]: message

File: lab_4_fill_words_by_ngrams/start.py [hunk 15]
```diff
@@ -3,14 +3,33 @@
...
     with open("./assets/Harry_Potter.txt", "r", encoding="utf-8") as text_file:
         text = text_file.read()
-    result = None
+    processor = WordProcessor('<eow>')
+    text_encoded = processor.encode(text)
+    model = NGramLanguageModel(text_encoded, 2)
+    model.build()
+    gen_top_p = TopPGenerator(model, processor, 0.5)
+    generation_1 = gen_top_p.run(51, 'Vernon')
+    print(generation_1)
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/start.py:23]: remove

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**